src/KurrentDB.SecondaryIndexing/Indexes/Diagnostics/SecondaryIndexProgressTracker.cs

```diff
@@ -0,0 +1,113 @@
+// Copyright (c) Kurrent, Inc and/or licensed to Kurrent, Inc under one or more agreements.
+// Kurrent, Inc licenses this file to you under the Kurrent License v1 (see LICENSE.md).
+
+using System.Diagnostics.Metrics;
+using KurrentDB.Core.Data;
+using Serilog;
+
+namespace KurrentDB.SecondaryIndexing.Indexes.Diagnostics;
+
+public interface ISecondaryIndexProgressTracker {
+	void RecordIndexed(ResolvedEvent resolvedEvent);
+	void RecordAppended(EventRecord eventRecord);
+	void RecordCommit(long position, int batchSize, long duration);
+	void RecordError(Exception e);
+}
+
+public class NoOpSecondaryIndexProgressTracker : ISecondaryIndexProgressTracker {
+	public void RecordIndexed(ResolvedEvent resolvedEvent) {
+	}
+
+	public void RecordAppended(EventRecord eventRecord) {
+	}
+
+	public void RecordCommit(long position, int batchSize, long duration) {
+	}
+
+	public void RecordError(Exception e) {
+	}
+}
+
+public class SecondaryIndexProgressTracker : ISecondaryIndexProgressTracker {
+	private static readonly ILogger Log = Serilog.Log.Logger.ForContext<SecondaryIndexProgressTracker>();
+	private long _lastIndexedPosition;
+	private long _lastCommittedPosition;
+	private long _lastLogPosition;
+	private long _lastIndexedAt;
+	private long _lastAppendedAt;
+	private int _pendingEvents;
+
+	public SecondaryIndexProgressTracker(Meter meter, string meterPrefix) {
+		meter.CreateObservableGauge(
+			$"{meterPrefix}.subscription.gap",
+			ObserveGap,
+			"events",
+			"Gap between last indexed and current last event log position"
+		);
+
+		meter.CreateObservableGauge(
+			$"{meterPrefix}.subscription.lag",
+			ObserveLag,
+			"s",
+			"Time between last appended and last indexed event"
+		);
+
+		meter.CreateObservableGauge(
+			$"{meterPrefix}.subscription.pending",
+			ObservePending,
+			"events",
+			"Events pending checkpoint"
+		);
+	}
+
+	public void RecordIndexed(ResolvedEvent resolvedEvent) {
+		try {
+			Interlocked.Exchange(ref _lastIndexedPosition, resolvedEvent.OriginalEvent.LogPosition);
+			Interlocked.Exchange(ref _lastIndexedAt, DateTime.UtcNow.Ticks);
+			Interlocked.Increment(ref _pendingEvents);
+		} catch (Exception exc) {
+			Log.Error(exc, "Error recording metrics of event indexed in secondary index.");
+		}
+	}
+
+	public void RecordAppended(EventRecord eventRecord) {
+		try {
+			Interlocked.Exchange(ref _lastLogPosition, eventRecord.LogPosition);
+			Interlocked.Exchange(ref _lastAppendedAt, eventRecord.TimeStamp.Ticks);
```

Comment:
there's no guarantee that the timestamp increases monotonically - for example if there's a leader change, some events in a chunk will have used the previous leader's clock while the new events will use the new leader's clock (both clocks should be quite close to each other though)

so, it might be worth adding a check to take the max of the current and new timestamps